### DISCUSSION TITLE
Timeouts on Promise

### DISCUSSION THREAD

**Discussion by FireMasterK** [activej/activej#158] [Q&A]
Hi there,

Is there any way to set a timeout for a `getResult()` call?

I would like to set a max timeout before which null, or an exception is thrown.

**Comment by FireMasterK** [comment]
Another question, how would I use promises with blocking calls?
```
        Promise<String> videoPromise = Promise.of(videoId);
        long start = System.currentTimeMillis();
        var p1 = videoPromise.map(id -> {
            Thread.sleep(1000);
            System.out.println(System.currentTimeMillis() - start);
            return "1" + id;
        });
        var p2 = videoPromise.map(id -> {
            Thread.sleep(2000);
            System.out.println(System.currentTimeMillis() - start);
            return "2" + id;
        });
```
This prints:
```
1000
3002
```

**Comment by eduard-vasinskyi** [comment]
Hi, @FireMasterK

`Promise#getResult` method is non-blocking. You cannot add a timeout to it. However you can timeout the whole promise by wrapping it into [Promises#timeout](https://github.com/activej/activej/blob/6d151e5bbfb7340e5baef07495e9741e3cf4ab75/core-promise/src/main/java/io/activej/promise/Promises.java#L61). For example:
```java
Promises.timeout(Duration.ofSeconds(5), longRunningTask())
```
If a promise  returned by the `longRunningTask()` method is not completed within 5 seconds the resulting promise (returned by `Promises.timeout(...)`) will be completed with an `AsyncTimeoutException`.

You should never block the eventloop thread. However, if you want to block on a promise (maybe from within some other thread) you can convert a promise to a `CompletableFuture` and then block waiting for the result of the `CompletableFuture`:
```java
longRunningTask().toCompletableFuture().get(5, TimeUnit.SECONDS)
```


>  how would I use promises with blocking calls?
As I said before you should not execute any blocking code within an eventloop thread. If you need to invoke a blocking call you need to use an external executor. For convenience, you can use `Promise#ofBlocking` methods which accept an executor alongside a blocking function. 
 
We can rewrite your snippet like this:
```java
Eventloop eventloop = Eventloop.create().withCurrentThread();
String videoId = "id";
 
ExecutorService executor = Executors.newCachedThreadPool();
Promise<String> videoPromise = Promise.of(videoId);
long start = System.currentTimeMillis();
var p1 = videoPromise
     .then(id -> Promise.ofBlocking(executor, () -> {
        Thread.sleep(1000);
        System.out.println(System.currentTimeMillis() - start);
        return "1" + id;
     }));
var p2 = videoPromise
     .then(id -> Promise.ofBlocking(executor, () -> {
        Thread.sleep(2000);
        System.out.println(System.currentTimeMillis() - start);
        return "2" + id;
     }));
 
eventloop.run();
executor.shutdown();

// prints 
// 1012
// 2011
 ```

You can also look at [UserDaoSql](https://github.com/activej/activej/blob/master/examples/core/sql/src/main/java/userservice/dao/UserDaoSql.java) class for example, to see how we use `Promise#ofBlocking` to transform blocking JDBC calls into asynchronous promises.

**Comment by FireMasterK** [comment]
Thank you, that was helpful!


However, there seems to be an issue when the blocking promise is put in a timeout:
```
        long start = System.currentTimeMillis();
        Eventloop eventloop = Eventloop.create().withCurrentThread();
        var prom = Promises.timeout(Duration.ofSeconds(1),
                Promise.ofBlocking(Multithreading.getCachedExecutor(), () -> {
                    Thread.sleep(2000);
                    return "test";
                })
        );
        eventloop.run();
        System.out.println(prom.getResult());
        System.out.println(System.currentTimeMillis() - start);
```

This prints:
```
null
2110
```

**Comment by eduard-vasinskyi** [comment]
The `prom` promise has timed out as the blocking code took more than 1 second to execute. The promise has been completed with a timeout exception instead of a result. That’s why you receive `null` as a result of a promise. If you call `prom.getException()` instead, you will receive an `AsyncTimeoutException`.

You can check whether a promise has already been completed using `Promise#isComplete` method. You can also use `Promise#isResult` and `Promise#isException` methods to determine whether a promise has been completed successfully or not.

However, it is not very common to use `getResult()`/`getException()` methods in an application code. Normally you would use `Promise#whenComplete`, `Promise#whenResult`, `Promise#whenException` methods to add callbacks to the promise execution. (note, that those callbacks would be executed inside the eventloop thread)
